src/py/flwr/server/driver/grpc_driver.py

```diff
@@ -140,56 +168,52 @@ class GrpcDriver(Driver):
+    run_id : int
+        The identifier of the run.
+    stub : Optional[GrpcDriverStub] (default: None)
+        The ``GrpcDriverStub`` instance used to communicate with the SuperLink.
+        If None, an instance connected to "[::]:9091" will be created.
     """
 
-    def __init__(
+    def __init__(  # pylint: disable=too-many-arguments
         self,
-        driver_service_address: str = DEFAULT_SERVER_ADDRESS_DRIVER,
-        root_certificates: Optional[bytes] = None,
-        fab_id: Optional[str] = None,
-        fab_version: Optional[str] = None,
+        run_id: int,
+        stub: Optional[GrpcDriverStub] = None,
     ) -> None:
-        self.addr = driver_service_address
-        self.root_certificates = root_certificates
-        self.driver_helper: Optional[GrpcDriverHelper] = None
-        self.run_id: Optional[int] = None
-        self.fab_id = fab_id if fab_id is not None else ""
-        self.fab_version = fab_version if fab_version is not None else ""
+        self._run_id = run_id
+        self._run: Optional[Run] = None
+        self.stub = stub if stub is not None else GrpcDriverStub()
         self.node = Node(node_id=0, anonymous=True)
 
-    def _get_grpc_driver_helper_and_run_id(self) -> Tuple[GrpcDriverHelper, int]:
-        # Check if the GrpcDriverHelper is initialized
-        if self.driver_helper is None or self.run_id is None:
-            # Connect and create run
-            self.driver_helper = GrpcDriverHelper(
-                driver_service_address=self.addr,
-                root_certificates=self.root_certificates,
+    @property
+    def run(self) -> Run:
+        """Run information."""
+        self._get_stub_and_run_id()
+        return Run(**vars(cast(Run, self._run)))
+
+    def _get_stub_and_run_id(self) -> Tuple[GrpcDriverStub, int]:
+        # Check if is initialized
+        if self._run is None:
+            # Connect
+            if not self.stub.is_connected():
+                self.stub.connect()
+            # Get the run info
+            req = GetRunRequest(run_id=self._run_id)
+            res = self.stub.get_run(req)
+            if not res.HasField("run"):
+                raise RuntimeError(f"Cannot find the run with ID: {self._run_id}")
+            self._run = Run(
+                run_id=res.run.run_id,
+                fab_id=res.run.fab_id,
+                fab_version=res.run.fab_version,
             )
-            self.driver_helper.connect()
-            req = CreateRunRequest(fab_id=self.fab_id, fab_version=self.fab_version)
-            res = self.driver_helper.create_run(req)
-            self.run_id = res.run_id
-        return self.driver_helper, self.run_id
+
+        return self.stub, self._run.run_id
 
     def _check_message(self, message: Message) -> None:
         # Check if the message is valid
         if not (
-            message.metadata.run_id == self.run_id
+            message.metadata.run_id == cast(Run, self._run).run_id
```

Comment:
Why is `cast` necessary here but not in `InMemoryDriver`?

Reply:
`InMemoryDriver` uses the property `self.run : Run` instead of the variable `self._run: Run | None`, and thus it doesn't need `cast`.